Add location and user profile to web user invite

stephherbers opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
tickets: [here](https://dimagi.atlassian.net/browse/USH-4375) and [here](https://dimagi.atlassian.net/browse/USH-4353) and additional [here](https://dimagi.atlassian.net/browse/USH-4509?focusedCommentId=326064&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-326064) for "assigned_locations" 


![Screen Shot 2024-05-01 at 5 38 56 PM](https://github.com/dimagi/commcare-hq/assets/88759246/c6de6177-3d0f-485d-ae02-8fae2cf69076)
![Screen Shot 2024-05-01 at 5 38 35 PM](https://github.com/dimagi/commcare-hq/assets/88759246/ed72342f-abe7-4df1-a583-e0803b9d85ea)

Also adds bulk upload compatible with "profiles" for users that require invites.

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
review as a whole
combined backend and frontend functionality.

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
This adds functionality that already exists via the bulk upload just to the initial invite rather than having to add in afterwards. This was QA'ed and tested locally and on staging. 

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
QA'ed and passed: https://dimagi.atlassian.net/browse/QA-6367

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

stephherbers committed:
ui changes: add profile and location dropdowns

stephherbers committed:
backend: add profile to web user

stephherbers committed:
Merge branch 'jt/web-user-invitation-model-setup-2' into smh/web-invite-update

stephherbers committed:
Merge branch 'master' into smh/web-invite-update

stephherbers committed:
Merge branch 'master' into smh/web-invite-update

stephherbers committed:
clean up last remains of supply_point reference

stephherbers committed:
Merge branch 'master' into smh/web-invite-update

stephherbers committed:
make profile not required field

stephherbers committed:
remove location_id from data before creating invite

Jtang-1 commented on corehq/apps/registration/forms.py:
```diff
@@ -487,24 +490,38 @@ class AdminInvitesUserForm(forms.Form):
+                    profiles = definition.get_profiles()
+                    profile_choices = [('', '')] + [(profile.id, profile.name) for profile in profiles]
+                self.fields['profile'].choices = profile_choices
+                show_profile = True
```
nit: could we set `show_profile`' directly as `domain_has_privilege(domain_obj.name, privileges.APP_USER_PROFILES)` and use it for the conditional check? And something similar for `show_location`

Jtang-1 replied:
no, that wouldn't work because it lives within the `domain_obj` conditional check. But another small clean up was done [764d0e4](https://github.com/dimagi/commcare-hq/pull/34466/commits/764d0e4a5145a20f98acb3e7ab1263e33bcc26f4)

Jtang-1 commented on corehq/apps/users/models.py:
```diff
@@ -518,14 +518,17 @@ def add_domain_membership(self, domain, timezone=None, **kwargs):
             self.set_location(domain, location_id)
+        if domain_has_privilege(domain_obj.name, privileges.APP_USER_PROFILES) and profile:
+            user_data = self.get_user_data(domain_obj.name)
+            user_data.update({}, profile_id=profile.id if profile.name else ...)
```
do we need the `{}`? And could you explain what `profile_id=profile.id if profile.name else ...` (in particular the Ellipsis) is doing? Could this just be `profile_id=profile.id` since we already checked that profile is not None?

Jtang-1 replied:
Ah, I thought `user_data` was a dict but it's an instance of `UserData` and `update` is one of its methods that requires the original `data` as an argument which is what the `{}` is for. `...` is also the default value so that checks out.
removed conditional check [bd74d28](https://github.com/dimagi/commcare-hq/pull/34466/commits/bd74d283922aea619cf85d24cdfe99b9e457c9de)

Jtang-1 commented on corehq/apps/users/views/__init__.py:
```diff
@@ -1143,6 +1143,12 @@ def post(self, request, *args, **kwargs):
                 data["domain"] = self.domain
+                location_id = data.pop("location_id", None)
+                data["location"] = SQLLocation.by_location_id(location_id) if location_id else None
+                profile_id = data.get("profile", None)
```
nit: should we also pop here

Jtang-1 commented on corehq/apps/registration/forms.py:
```diff
@@ -487,24 +490,38 @@ class AdminInvitesUserForm(forms.Form):
     email = forms.EmailField(label="Email Address",
                              max_length=User._meta.get_field('email').max_length)
     role = forms.ChoiceField(choices=(), label="Project Role")
+    profile = forms.ChoiceField(choices=(), label="Profile", required=False)
```
nit: just throwing this out, would the variable name `profile_id` be more clear since that's what's stored? It potentially could be more confusing, so i'm fine with leaving it as is too.

Jtang-1 replied:
going to leave it as is

orangejenny committed:
Merge branch 'master' into smh/web-invite-update

Jtang-1 committed:
refactor to remove flags

Jtang-1 committed:
removes choices inititialization, choices already defaults as empty list

Jtang-1 committed:
conditional check for profile is already done so does not need to be repeated

Jtang-1 committed:
specify id for widget and remove unused initial value

Jtang-1 commented on corehq/apps/registration/forms.py:
```diff
@@ -487,24 +490,38 @@ class AdminInvitesUserForm(forms.Form):
-            self.fields['program'].choices = choices
+                                                         widget=LocationSelectWidget(domain_obj.name),
+                                                         help_text=EMWF.location_search_help,
+                                                         initial='')
```
`initial` wasn't being used, it's controlled by the widget

Jtang-1 committed:
lint - remove duplicate import

Jtang-1 commented:
I carried this PR over from Steph and implemented the comments I had. Looking for an additional person's review/approval - PR is ready for review.

orangejenny commented on corehq/apps/registration/forms.py:
```diff
@@ -493,18 +496,28 @@ def __init__(self, data=None, excluded_emails=None, is_add_user=None, location=N
         domain_obj = Domain.get_by_name(domain)
         self.fields['role'].choices = role_choices
-        if domain_obj.commtrack_enabled:
+        if domain_obj:
```
Isn't locations privilege-restricted?

It also seems odd to me that this would *always* be visible, as I don't think all projects use locations. Maybe it's worth checking to see if the project has any locations defined?

Jtang-1 replied:
agreed [b83869f](https://github.com/dimagi/commcare-hq/pull/34466/commits/b83869f47a02da39ac65cb7c33913ff107e2f1ff)

orangejenny commented on corehq/apps/registration/forms.py:
```diff
@@ -493,18 +496,28 @@ def __init__(self, data=None, excluded_emails=None, is_add_user=None, location=N
             self.fields['location_id'] = forms.CharField(label='Primary Location', required=False,
-                                                          widget=LocationSelectWidget(domain_obj.name),
-                                                          help_text=EMWF.location_search_help,
-                                                          initial=location.location_id if location else '')
```
Why was the initial value removed?

orangejenny replied:
nevermind, I see this discussed in the resolve comments

Jtang-1 replied:
Did some more digging and this location seemed to have been added as part of the `ewsghana` project in this [PR](https://github.com/dimagi/commcare-hq/commit/2b74cf7b47fca54ada43ad44f6a259bc7c659ce5#diff-2912c4655bb3c474019c855a831566e13abfe17707fd9fc7b316118de4a233a3). It has since all been deleted but this additions related to the web user invite field and form seemed to have been missed.

orangejenny commented on corehq/apps/registration/forms.py:
```diff
@@ -493,18 +496,28 @@ def __init__(self, data=None, excluded_emails=None, is_add_user=None, location=N
+                                                                                    id='id_location_id'),
+                                                         help_text=EMWF.location_search_help)
+
+            if domain_has_privilege(domain_obj.name, privileges.APP_USER_PROFILES):
```
I think it'd also be nice to only show this is there are profiles defined.

Jtang-1 replied:
[36ab85d](https://github.com/dimagi/commcare-hq/pull/34466/commits/36ab85df23b87dd6db30e21d3b648721794f0d34)

orangejenny commented on corehq/apps/registration/forms.py:
```diff
@@ -521,6 +534,8 @@ def __init__(self, data=None, excluded_emails=None, is_add_user=None, location=N
                     data_bind="textInput: email",
                 ),
                 'role',
+                'profile' if 'profile' in self.fields else None,
```
How were these being displayed before? And why doesn't `program` need to be in here?

Jtang-1 replied:
I found that strange too, i don't know how `supply_point` / `location` and `program` are ever shown. But I do see invites created in the past 3 months with non-null values for those fields. Possibly those were all from bulk upload? 

The layout was added in [this commit ](https://github.com/dimagi/commcare-hq/commit/e1c619f2d67ce2bedc0b2adae720f6e9023acd0c)3 years ago so, I decided to just leave it as is for those fields and not mess with it within the scope of this PR.

orangejenny replied:
Makes sense to me.

orangejenny reviewed (commented):
A couple of comments/questions. My crispy forms knowledge is not strong.

Jtang-1 committed:
additional cleanup of location arguments

These were added and used in ewsghana files https://github.com/dimagi/commcare-hq/commit/2b74cf7b47fca54ada43ad44f6a259bc7c659ce5#diff-7a351958b4a439f872148d81677d07c31afe87e377fcdc4052e34712f6888c9d
Those files have since been deleted and there doesn't exist any other calls to the
view that passes location_id

Jtang-1 committed:
show locations in form only if locations are used in project

Jtang-1 committed:
show profiles UI field only if profiles are defined

Jtang-1 committed:
Merge branch 'master' into smh/web-invite-update

Jtang-1 committed:
rename field to specify it is the primary location

Jtang-1 committed:
correct import

Jtang-1 committed:
shows location on post request to invite page

this occurs if there's an error upon sending invite and the
invite web user view is called again

orangejenny commented on corehq/apps/users/views/__init__.py:
```diff
@@ -1095,6 +1095,7 @@ def invite_web_user_form(self):
             role_choices=role_choices,
             domain=self.domain,
             is_add_user=is_add_user,
+            should_show_location=self.request.project.uses_locations
```
It's convenient that this already exists.

orangejenny commented on corehq/apps/registration/forms.py:
```diff
@@ -491,7 +491,7 @@ class AdminInvitesUserForm(forms.Form):
                              max_length=User._meta.get_field('email').max_length)
     role = forms.ChoiceField(choices=(), label="Project Role")
 
-    def __init__(self, data=None, excluded_emails=None, is_add_user=None, location=None,
```
Nice cleanup, and thank you for the explanation in the commit description.

orangejenny commented:
Looks like there's an error in the most recent master merge. Everything else looks good.

Jtang-1 committed:
extract form for selecting assigned and primary locations for user

orangejenny commented on corehq/apps/users/views/__init__.py:
```diff
@@ -1081,6 +1081,7 @@ def invite_web_user_form(self):
                 role_choices=role_choices,
                 domain=self.domain,
                 is_add_user=is_add_user,
+                should_show_location=self.request.project.uses_locations
```
appreciate the commit comment

Jtang-1 committed:
use inherited baseLocationForm class for DRY

Jtang-1 committed:
these field names will alwas exist within self.fields

Jtang-1 commented:
We recently discovered that "assigned_locations" also need to be included in the invitation workflow. So the past 2 commits and there will be additional upcoming commits to implement that. Also will need to align with some additional model changes.

orangejenny commented:
@Jtang-1 👍 , ping when you're ready for re-review.

Jtang-1 committed:
Merge branch 'rc/assigned-locations-invite-migration' into smh/web-invite-update

Jtang-1 committed:
fetch data according to new primary_location form field name

Jtang-1 committed:
fetch assign_locations from form to create Invitation

Jtang-1 committed:
add missing profile argument when creating web user via management command

Jtang-1 committed:
Merge branch 'rc/assigned-locations-invite-migration' into smh/web-invite-update

Jtang-1 committed:
bug: many to many field assignments must be done with .set() or .add()

the invitation must be created first before the field can get set

Jtang-1 committed:
Merge branch 'rc/assigned-locations-invite-migration' into smh/web-invite-update

Jtang-1 committed:
Merge branch 'rc/assigned-locations-invite-migration' into smh/web-invite-update

Jtang-1 commented:
ready for re-review!

Jtang-1 committed:
Merge branch 'rc/assigned-locations-invite-migration' into smh/web-invite-update

Jtang-1 committed:
Merge branch 'rc/assigned-locations-invite-migration' into smh/web-invite-update

Jtang-1 committed:
feat: support profiles when bulk uploading new users to domain and creating invite

orangejenny commented:
I probably won't get to reviewing this today but should get there tomorrow.

orangejenny commented on corehq/apps/users/views/__init__.py:
```diff
@@ -1136,6 +1138,10 @@ def post(self, request, *args, **kwargs):
                                         if primary_location_id else None)
+                assigned_location_ids = data.pop("assigned_locations", None)
+                data["assigned_locations"] = [SQLLocation.by_location_id(assigned_location_id)
+                                        if assigned_location_id else None
```
Are these ids are ever expected to be falsy, or is this just defensive programming?

Jtang-1 replied:
they're not expected to be false so just defensively programming. But I really do just want to avoid any "None" values [6816b2f](https://github.com/dimagi/commcare-hq/pull/34466/commits/6816b2f0053a331465c4ce141129933d1667da6f)

orangejenny commented on corehq/apps/user_importer/importer.py:
```diff
@@ -653,10 +655,14 @@ def _process_web_user(self):
                                                 self.user.assigned_location_ids)
                 elif not web_user or not web_user.is_member_of(self.domain):
+                    profile = None
+                    if cv["profile_name"] in self.domain_info.profiles_by_name:
```
Should this report an error/warning if the profile isn't found?

Jtang-1 replied:
that would be a nice feedback to have [f1b723c](https://github.com/dimagi/commcare-hq/pull/34466/commits/f1b723c73d2878e52a47d4f6fe8cf141abd8c5ca)

orangejenny replied:
beautiful

orangejenny reviewed (approved):
Looks good, left a couple minor questions.

Jtang-1 committed:
the model will fail when setting the field to a list containing None

assigned_location_id is not expected to be None, but defensively check that here

Jtang-1 committed:
raise error if profile name is not valid

orangejenny commented on corehq/apps/user_importer/importer.py:
```diff
@@ -727,10 +733,14 @@ def process_existing_user(self, user):
                 )
             else:
+                profile = None
+                if self.column_values["profile_name"] in self.domain_info.profiles_by_name:
```
Does this one also need user feedback if the profile isn't found?

Jtang-1 replied:
yes it does [34215de](https://github.com/dimagi/commcare-hq/pull/34466/commits/34215de14cfdea0340b31f3b2b860ee0b7491a5c)

Jtang-1 committed:
refactor out profile validation and add missed profile validation for processing existing user

Jtang-1 committed:
Merge branch 'rc/assigned-locations-invite-migration' into smh/web-invite-update
